x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Intersects.java

```diff
@@ -0,0 +1,234 @@
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.TypeResolutions;
+import org.elasticsearch.xpack.ql.expression.function.scalar.BinaryScalarFunction;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+import org.elasticsearch.xpack.ql.util.SpatialUtils;
+
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isType;
+import static org.elasticsearch.xpack.ql.type.DataTypes.GEO_POINT;
+import static org.elasticsearch.xpack.ql.type.DataTypes.isString;
+
+public class Intersects extends BinaryScalarFunction implements Mappable {
+
+    public Intersects(Source source, Expression timestamp, Expression argument) {
+        super(source, timestamp, argument);
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataTypes.BOOLEAN;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        TypeResolution resolution = isGeoPointOrString(left(), sourceText(), FIRST);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        resolution = isGeoPointOrString(right(), sourceText(), SECOND);
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+        return TypeResolution.TYPE_RESOLVED;
+    }
+
+    public static TypeResolution isGeoPointOrString(Expression e, String operationName, TypeResolutions.ParamOrdinal paramOrd) {
+        return isType(e, dt -> dt == GEO_POINT || isString(dt), operationName, paramOrd, "geo_point", "keyword", "text");
+    }
+
+    @Override
+    protected Intersects replaceChildren(Expression newLeft, Expression newRight) {
+        return new Intersects(source(), newLeft, newRight);
+    }
+
+    @Override
+    protected NodeInfo<? extends Expression> info() {
+        return NodeInfo.create(this, Intersects::new, left(), right());
+    }
+
+    @Override
+    public boolean foldable() {
+        return left().foldable() && right().foldable();
+    }
+
+    @Override
+    public Object fold() {
+        return Mappable.super.fold();
+    }
+
+    @Override
+    public Supplier<EvalOperator.ExpressionEvaluator> toEvaluator(
+        Function<Expression, Supplier<EvalOperator.ExpressionEvaluator>> toEvaluator
+    ) {
+        if (isString(left().dataType())) {
+            if (isString(right().dataType())) {
+                if (left().foldable()) {
+                    if (right().foldable()) {
+                        // Both are constant strings
+                        return () -> new IntersectsConstantConstantEvaluator(asString(left()), asString(right()));
```

Comment:
Ah. I see why `ConstantConstant` name. You don't actually have to optimize this case because we'll catch it at planning time and call `Mappable.super.fold()` above. So long as the variable string representation versions *work* it isn't super important how fast this is because constant folding is out of the hot path.